File: src/apprt/gtk/BroadcastGroups.zig
//! Manages "broadcast input" groups: sets of terminal surfaces that all
//! receive the keyboard input typed into any one of them, similar to
//! iTerm2's broadcast input and tmux's synchronized panes.
//!
//! Surfaces are toggled in and out of groups with ctrl+shift+click, or
//! join a specific numbered group with the `join_broadcast_group`
//! keybinding action (ctrl+shift+<digit> by default). Multiple groups can
//! exist at once and each group is assigned a distinct border color so
//! its members are visually identifiable. At most `max_groups` groups can
//! exist, one per palette color.
//!
//! Surfaces are tracked by their core surface ID so that this structure
//! never holds a pointer that could dangle; members are removed explicitly
//! when a surface is destroyed.
const BroadcastGroups = @This();

const std = @import("std");
const Allocator = std.mem.Allocator;

/// The number of distinct border colors available for groups. This must
/// match the number of `broadcast-color-N` CSS classes defined in
/// css/style.css.
pub const color_count = 10;

/// The maximum number of groups that can exist at once. Each group owns
/// one palette color, so this is bounded by the palette size.
pub const max_groups = color_count;

const Group = struct {
    /// The number of this group, which doubles as its palette color
    /// index. Always less than `max_groups`. Click-created groups take
    /// the lowest number unused by any other group; the keyboard
    /// shortcut addresses groups by this number directly.
    color: u8,

    /// The core surface IDs of the group members.
    members: std.ArrayListUnmanaged(u64) = .empty,
};

groups: std.ArrayListUnmanaged(Group) = .empty,

pub const empty: BroadcastGroups = .{};

pub fn deinit(self: *BroadcastGroups, alloc: Allocator) void {
    for (self.groups.items) |*group| group.members.deinit(alloc);
    self.groups.deinit(alloc);
}

/// Toggle the given surface's group membership: a surface already in a
/// group is removed from it (empty groups dissolve); otherwise, if the
/// focused surface is in a group then the given surface joins that group;
/// otherwise a new group is created containing only the given surface.
/// If all `max_groups` groups already exist, no new group is created and
/// this is a no-op.
pub fn toggle(
    self: *BroadcastGroups,
    alloc: Allocator,
    id: u64,
    focused_id: ?u64,
) Allocator.Error!void {
    // Already a member somewhere: remove.
    if (self.groupIndex(id)) |idx| {
        self.removeFromGroup(alloc, idx, id);
        return;
    }

    // If the focused surface is in a group, join that group.
    if (focused_id) |focused| {
        if (self.groupIndex(focused)) |idx| {
            try self.groups.items[idx].members.append(alloc, id);
            return;
        }
    }

    // Start a new group, unless every group number is already taken.
    const color = self.nextColor() orelse return;
    var group: Group = .{ .color = color };
    errdefer group.members.deinit(alloc);
    try group.members.append(alloc, id);
    try self.groups.append(alloc, group);
}

/// Add the given surface to the group with the given number, creating
/// the group if it doesn't exist yet. A surface already in a different
/// group is moved; a surface already in that exact group is removed
/// from it instead (toggle). The number must be less than `max_groups`.
pub fn join(
    self: *BroadcastGroups,
    alloc: Allocator,
    id: u64,
    number: u8,
) Allocator.Error!void {
    std.debug.assert(number < max_groups);

    // Leave the current group, if any. If it was the requested group
    // then this is a toggle off and we're done.
    if (self.groupIndex(id)) |idx| {
        const current = self.groups.items[idx].color;
        self.removeFromGroup(alloc, idx, id);
        if (current == number) return;
    }

    // Join the requested group, creating it if needed. Note that
    // removeFromGroup above may have reordered `groups` so we search
    // by number only now.
    const idx = self.groupIndexByColor(number) orelse idx: {
        try self.groups.append(alloc, .{ .color = number });
        break :idx self.groups.items.len - 1;
    };
    try self.groups.items[idx].members.append(alloc, id);
}

/// Remove the given surface from its group, if any. Called when a
/// surface is destroyed.
pub fn remove(self: *BroadcastGroups, alloc: Allocator, id: u64) void {
    const idx = self.groupIndex(id) orelse return;
    self.removeFromGroup(alloc, idx, id);
}

/// The color assigned to the given surface's group, or null if the
/// surface is not in any group.
pub fn colorOf(self: *const BroadcastGroups, id: u64) ?u8 {
    const idx = self.groupIndex(id) orelse return null;
    return self.groups.items[idx].color;
}

/// All members of the given surface's group (including the surface
/// itself), or null if the surface is not in any group. The returned
/// slice is invalidated by any mutation of this structure.
pub fn members(self: *const BroadcastGroups, id: u64) ?[]const u64 {
    const idx = self.groupIndex(id) orelse return null;
    return self.groups.items[idx].members.items;
}

fn groupIndexByColor(self: *const BroadcastGroups, color: u8) ?usize {
    for (self.groups.items, 0..) |group, i| {
        if (group.color == color) return i;
    }

    return null;
}

fn groupIndex(self: *const BroadcastGroups, id: u64) ?usize {
    for (self.groups.items, 0..) |group, i| {
        for (group.members.items) |member| {
            if (member == id) return i;
        }
    }

    return null;
}

fn removeFromGroup(
    self: *BroadcastGroups,
    alloc: Allocator,
    idx: usize,
    id: u64,
) void {
    const group = &self.groups.items[idx];
    for (group.members.items, 0..) |member, i| {
        if (member == id) {
            _ = group.members.swapRemove(i);
            break;
        }
    }

    // Dissolve empty groups so their color becomes available again.
    if (group.members.items.len == 0) {
        group.members.deinit(alloc);
        _ = self.groups.swapRemove(idx);
    }
}

/// The lowest group number not used by any existing group, or null if
/// all `max_groups` numbers are taken.
fn nextColor(self: *const BroadcastGroups) ?u8 {
    var candidate: u8 = 0;
    while (candidate < max_groups) : (candidate += 1) {
        if (self.groupIndexByColor(candidate) == null) return candidate;
    }

    return null;
}

test "toggle creates, joins, and removes" {
    const testing = std.testing;
    const alloc = testing.allocator;

    var groups: BroadcastGroups = .empty;
    defer groups.deinit(alloc);

    // New group with the clicked surface.
    try groups.toggle(alloc, 1, null);
    try testing.expectEqual(@as(?u8, 0), groups.colorOf(1));

    // Focused surface in a group: join it.
    try groups.toggle(alloc, 2, 1);
    try testing.expectEqual(@as(?u8, 0), groups.colorOf(2));
    try testing.expectEqual(@as(usize, 2), groups.members(1).?.len);

    // Focused surface not in a group: new group with a new color.
    try groups.toggle(alloc, 3, null);
    try testing.expectEqual(@as(?u8, 1), groups.colorOf(3));

    // Toggling a member removes it.
    try groups.toggle(alloc, 2, 1);
    try testing.expectEqual(@as(?u8, null), groups.colorOf(2));

    // Removing the last member dissolves the group and frees its color.
    try groups.toggle(alloc, 1, null);
    try testing.expectEqual(@as(?u8, null), groups.colorOf(1));
    try groups.toggle(alloc, 4, null);
    try testing.expectEqual(@as(?u8, 0), groups.colorOf(4));
}

test "toggle refuses to create more than max_groups groups" {
    const testing = std.testing;
    const alloc = testing.allocator;

    var groups: BroadcastGroups = .empty;
    defer groups.deinit(alloc);

    // Fill every group number with a single-member group.
    var id: u64 = 1;
    while (id <= max_groups) : (id += 1) {
        try groups.toggle(alloc, id, null);
        try testing.expectEqual(@as(?u8, @intCast(id - 1)), groups.colorOf(id));
    }

    // The next toggle would need an eleventh group: it must be a no-op.
    try groups.toggle(alloc, 99, null);
    try testing.expectEqual(@as(?u8, null), groups.colorOf(99));

    // Joining an existing group is still allowed while full.
    try groups.toggle(alloc, 99, 1);
    try testing.expectEqual(@as(?u8, 0), groups.colorOf(99));
}

test "join creates, moves, and toggles" {
    const testing = std.testing;
    const alloc = testing.allocator;

    var groups: BroadcastGroups = .empty;
    defer groups.deinit(alloc);

    // Joining a group that doesn't exist creates it with that number.
    try groups.join(alloc, 1, 4);
    try testing.expectEqual(@as(?u8, 4), groups.colorOf(1));

    // Another surface joins the same group.
    try groups.join(alloc, 2, 4);
    try testing.expectEqual(@as(?u8, 4), groups.colorOf(2));
    try testing.expectEqual(@as(usize, 2), groups.members(1).?.len);

    // Joining a different group moves the surface.
    try groups.join(alloc, 2, 7);
    try testing.expectEqual(@as(?u8, 7), groups.colorOf(2));
    try testing.expectEqual(@as(usize, 1), groups.members(1).?.len);

    // Joining the surface's current group removes it (toggle off), and
    // the emptied group dissolves so its number is free again.
    try groups.join(alloc, 2, 7);
    try testing.expectEqual(@as(?u8, null), groups.colorOf(2));
    try groups.toggle(alloc, 3, null);
    try testing.expectEqual(@as(?u8, 0), groups.colorOf(3));

    // The highest valid group number works.
    try groups.join(alloc, 5, max_groups - 1);
    try testing.expectEqual(@as(?u8, max_groups - 1), groups.colorOf(5));
}

test "remove" {
    const testing = std.testing;
    const alloc = testing.allocator;

    var groups: BroadcastGroups = .empty;
    defer groups.deinit(alloc);

    try groups.toggle(alloc, 1, null);
    try groups.toggle(alloc, 2, 1);
    groups.remove(alloc, 1);
    try testing.expectEqual(@as(?u8, null), groups.colorOf(1));
    try testing.expectEqual(@as(?u8, 0), groups.colorOf(2));

    // Removing a surface that isn't in a group is a no-op.
    groups.remove(alloc, 99);
}

File: macos/Sources/Features/Terminal/BroadcastGroups.swift
import AppKit
import SwiftUI

/// Manages "broadcast input" groups: sets of terminal surfaces that all receive the
/// same keyboard input, similar to iTerm2's broadcast input and tmux's synchronized
/// panes.
///
/// Surfaces are toggled in and out of groups with cmd+shift+click, or join a
/// specific numbered group with the `join_broadcast_group` keybinding action
/// (cmd+ctrl+<digit> by default). Multiple groups can exist at
/// once and each group is assigned a distinct border color so members are
/// visually identifiable. At most `maxGroups` groups can exist, one per palette
/// color.
///
/// Membership rules for a cmd+shift+click on a surface:
///   - If the surface is already in a group, it is removed from that group.
///   - Otherwise, if the currently focused surface is in a group, the clicked
///     surface joins that group.
///   - Otherwise, a new group is created containing the clicked surface,
///     unless all group numbers are already in use.
///
/// Members are held weakly so closed surfaces fall out of their group naturally.
@MainActor
class BroadcastGroups {
    static let shared = BroadcastGroups()

    /// Border colors assigned to groups. A group's number is its index into
    /// this palette.
    static let palette: [Color] = [
        .orange, .cyan, .green, .purple, .yellow,
        .pink, .red, .blue, .brown, .gray,
    ]

    /// The maximum number of groups that can exist at once. Each group owns
    /// one palette color, so this is bounded by the palette size.
    static let maxGroups = palette.count

    private struct Group {
        /// The number of this group, which doubles as its palette index.
        /// Always less than `maxGroups`.
        let number: Int

        var members: [Weak<Ghostty.SurfaceView>]

        var liveMembers: [Ghostty.SurfaceView] {
            members.compactMap { $0.value }
        }
    }

    private var groups: [Group] = []

    private init() {}

    /// Toggle the group membership of the given surface. See the class docs for
    /// the membership rules.
    func toggle(_ surfaceView: Ghostty.SurfaceView) {
        prune()

        // Already in a group: remove it.
        if let idx = groupIndex(of: surfaceView) {
            groups[idx].members.removeAll { $0.value === surfaceView }
            surfaceView.broadcastGroupColor = nil
            if groups[idx].members.isEmpty {
                groups.remove(at: idx)
            }
            return
        }

        // Join the focused surface's group if it has one, else start a new group.
        let targetIdx: Int
        if let focused = Self.focusedSurface,
           let focusedIdx = groupIndex(of: focused) {
            targetIdx = focusedIdx
        } else {
            // Take the lowest unused group number. If every number is in
            // use we refuse: there can be at most maxGroups groups.
            let used = Set(groups.map { $0.number })
            guard let number = (0..<Self.maxGroups).first(where: { !used.contains($0) }) else {
                return
            }
            groups.append(Group(number: number, members: []))
            targetIdx = groups.count - 1
        }

        add(surfaceView, toGroupAt: targetIdx)
    }

    /// Add the given surface to the group with the given number (0-based),
    /// creating the group if it doesn't exist yet. A surface already in a
    /// different group is moved; a surface already in that exact group is
    /// removed from it instead (toggle). Returns false if the number is
    /// outside 0..<maxGroups.
    @discardableResult
    func join(_ surfaceView: Ghostty.SurfaceView, group number: Int) -> Bool {
        guard (0..<Self.maxGroups).contains(number) else { return false }
        prune()

        // Leave the current group, if any. If it was the requested group
        // then this is a toggle off and we're done.
        if let idx = groupIndex(of: surfaceView) {
            let current = groups[idx].number
            groups[idx].members.removeAll { $0.value === surfaceView }
            surfaceView.broadcastGroupColor = nil
            if groups[idx].members.isEmpty {
                groups.remove(at: idx)
            }
            if current == number { return true }
        }

        // Join the requested group, creating it if needed.
        let targetIdx: Int
        if let idx = groups.firstIndex(where: { $0.number == number }) {
            targetIdx = idx
        } else {
            groups.append(Group(number: number, members: []))
            targetIdx = groups.count - 1
        }

        add(surfaceView, toGroupAt: targetIdx)
        return true
    }

    private func add(_ surfaceView: Ghostty.SurfaceView, toGroupAt idx: Int) {
        groups[idx].members.append(Weak(surfaceView))
        surfaceView.broadcastGroupColor = Self.palette[groups[idx].number]
    }

    /// Returns all other live members of the given surface's group. Empty if the
    /// surface is not in a group or is the only member. Members not attached to
    /// a window (e.g. closed surfaces kept alive by the undo manager) are
    /// excluded so input never reaches an invisible terminal.
    func others(for surfaceView: Ghostty.SurfaceView) -> [Ghostty.SurfaceView] {
        guard let idx = groupIndex(of: surfaceView) else { return [] }
        return groups[idx].liveMembers.filter { $0 !== surfaceView && $0.window != nil }
    }

    private func groupIndex(of surfaceView: Ghostty.SurfaceView) -> Int? {
        groups.firstIndex { group in
            group.members.contains { $0.value === surfaceView }
        }
    }

    /// Drop deallocated members and dissolve empty groups.
    private func prune() {
        for idx in groups.indices {
            groups[idx].members.removeAll { $0.value == nil }
        }
        groups.removeAll { $0.members.isEmpty }
    }

    /// The surface that currently has keyboard focus, if any.
    private static var focusedSurface: Ghostty.SurfaceView? {
        (NSApp.keyWindow?.windowController as? BaseTerminalController)?.focusedSurface
    }
}
